pkg/asset/quota/openstack/openstack.go

```diff
@@ -11,44 +14,90 @@ import (
 // These constraints can be used to check if there is enough quota for creating a cluster
 // for the install config.
 func Constraints(ci *validation.CloudInfo, controlPlanes []machineapi.Machine, computes []machineapi.MachineSet) []quota.Constraint {
-	ctrplConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(controlPlanes))
-	for i, m := range controlPlanes {
-		ctrplConfigs[i] = m.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	var constraints []quota.Constraint
+
+	for i := 0; i < len(controlPlanes); i++ {
+		constraints = append(constraints, machineConstraints(ci, &controlPlanes[i])...)
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(controlPlanes))))
+
+	for i := 0; i < len(computes); i++ {
+		constraints = append(constraints, machineSetConstraints(ci, &computes[i])...)
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(computes))))
+
+	return aggregate(constraints)
+}
+
+func getOpenstackProviderSpec(spec *machineapi.ProviderSpec) *openstackprovider.OpenstackProviderSpec {
+	// An invalid ProviderSpec would presumably be a programming error here.
+	if spec.Value == nil {
+		logrus.Warnf("Empty ProviderSpec")
+		return nil
 	}
-	computeReplicas := make([]int64, len(computes))
-	computeConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(computes))
-	for i, m := range computes {
-		computeReplicas[i] = int64(*m.Spec.Replicas)
-		computeConfigs[i] = m.Spec.Template.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+
+	osps, ok := spec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	if !ok {
+		logrus.Warnf("ProviderSpec of unexpected type %T", spec.Value.Object)
```

Comment:
Honestly I'd be fine with panicking here: if the type assertion fails, then the panic makes it very clear that there is a programming error somewhere.

Reply:
Panic with a specific error message (i.e. turn this into Errorf), or just remove this checking entirely and leave it to fail on its own?

Reply:
I’d really leave it to fail on its own, ie the good ol’ type assertion panic.

Reply:
Agree. I've left the spec.Value check in so the 2 different errors can be distinguished, but I've also turned it into a panic and removed further error checking around it.